src/com/amazon/ion/impl/bin/IonRawBinaryWriter.java

```diff
@@ -847,8 +847,10 @@ private void prepareValue()
             final long annotationsLengthPosition = buffer.position();
             buffer.writeVarUInt(0L);
             int annotationsLength = 0;
-            for (final int symbol : currentAnnotationSids)
+            // XXX: This is a very hot path. This code intentionally avoids creating iterators.
+            for (int m = 0; m < currentAnnotationSids.size(); m++)
             {
+                final int symbol = currentAnnotationSids.get(m);
```

Comment:
This makes me happy and frustrated at the same time. Happy for the win, frustrated that Java requires this syntactic difference to get it. 

Well done.

Reply:
This bit was a pretty conservative change. 

When I ran the code in a memory profiler, it reported a large number of allocations happening here. However, when I ran the code in JMH without this change (and with profiling disabled), the difference I'd see in allocations was minimal.

I've read that JVM memory profilers have to instrument the byte code for counting and that can change it just enough that it confounds the JIT; I suspect that's what's happening here. When running normally, the JIT can probably boil these down to the same thing. When profiling, though, it becomes a hotspot. A real heisenbug.

The same cannot be said for the other two changes, however! I could reliably see improvements from those no matter how I ran the code.